### PR TITLE
new clock implementation and tests

### PR DESCRIPTION
**Pull request by stephenctw**
- Implement real clock
- Achieve 100% function test coverage, excluding `machine` related ones
- Add `forge fmt --check` and `forge test` to CI
- Update lua node to keep track of block time after fastforwarding
- (lua-node) Add two idle players, which will be eliminated by honest player calling `eliminateMatchByTimeout`

### CODE REVIEW

**Comment by stephenctw** [comment]
I failed to re-enable `Match.t.sol` because of this [issue](https://github.com/foundry-rs/foundry/issues/432).

**Comment by stephenctw** [comment]
@GCdePaula I made some changes based on our discussion awhile ago. But I don't think I've covered everything. Could you submit your comments/reviews so I can double check?

**Review by GCdePaula — COMMENTED** [review]
Amazing work! I've left a few comments.

File: permissionless-arbitration/contracts/src/CanonicalConstants.sol [hunk 1]
```diff
@@ -10,41 +10,37 @@ library ArbitrationConstants {
...
-            Time.Duration.unwrap(CENSORSHIP_TOLERANCE) +
-                Time.Duration.unwrap(VALIDATOR_EFFORT)
-        );
+    // Dummy
+    Time.Duration constant COMMITMENT_EFFORT = Time.Duration.wrap(0);
+    Time.Duration constant CENSORSHIP_TOLERANCE = Time.Duration.wrap(60 * 5);
+    Time.Duration constant MATCH_EFFORT = Time.Duration.wrap(60 * 2);
 
     // 4-level tournament
     uint64 constant LEVELS = 3;
```
[the remark this answers is not in the document]
Comment (stephenctw) [reply permissionless-arbitration/contracts/src/CanonicalConstants.sol:32]: What is it supposed to be?

File: permissionless-arbitration/contracts/src/tournament/abstracts/LeafTournament.sol [hunk 2]
```diff
@@ -73,78 +73,63 @@ abstract contract LeafTournament is Tournament {
...
 
         if (_leftNode.join(_rightNode).eq(_matchId.commitmentOne)) {
             require(
-                _finalState.eq(_finalStateOne),
-                "final state one doesn't match"
+                _finalState.eq(_finalStateOne), "final state one doesn't match"
             );
 
-            _clockOne.addValidatorEffort(Time.ZERO_DURATION);
+            _clockOne.deduct(_clockTwo.timeSinceTimeout());
```
[the remark this answers is not in the document]
Comment (stephenctw) [reply permissionless-arbitration/contracts/src/tournament/abstracts/LeafTournament.sol:83]: I see, so a simple `setPaused()` should do it here.

[the remark this answers is not in the document]
Comment (stephenctw) [reply permissionless-arbitration/contracts/src/tournament/abstracts/LeafTournament.sol:83]: I see, so a simple `setPaused()` should do it here.

**Review by GCdePaula — APPROVED** [review]
Great work :)

File: permissionless-arbitration/contracts/src/tournament/abstracts/NonRootTournament.sol [hunk 3]
```diff
@@ -45,37 +44,41 @@ abstract contract NonRootTournament is Tournament {
     }
 
     /// @notice get the dangling commitment at current level and then retrieve the winner commitment
-    function innerTournamentWinner() external view returns (bool, Tree.Node) {
+    /// @return (bool, Tree.Node, Tree.Node)
+    /// - if the tournament is finished
+    /// - the contested parent commitment
+    /// - the dangling commitment
+    function innerTournamentWinner()
```
[the remark this answers is not in the document]
Comment (stephenctw) [reply permissionless-arbitration/contracts/src/tournament/abstracts/NonRootTournament.sol:52]: https://github.com/cartesi/dave/issues/7

**Pull request merged**